va/caa.go

```diff
@@ -160,7 +143,29 @@ func (va *ValidationAuthorityImpl) parallelCAALookup(ctx context.Context, name s
 	return results
 }
 
-func (va *ValidationAuthorityImpl) getCAASet(ctx context.Context, hostname string) (*CAASet, string, error) {
+// selectCAA picks the relevant CAA resource record set to be used, i.e. the
+// set for the "closest parent" of the FQDN in question, including the domain
+// itself. If the first non-empty RR set we find encountered an error during
+// lookup, we must assume that there could have been real records hidden by the
+// error, and propagate that error upwards.
+func selectCAA(rrs []caaResult) (*caaResult, error) {
+	for _, res := range rrs {
+		if res.err != nil {
+			return nil, res.err
+		}
+		if len(res.raw) > 0 {
```

Comment:
Previously, we would terminate this loop `if len(res.records) > 0 {`, in other words if there were non-zero numbers of records returned. Here we terminate if the dig-style string representation is non-empty. I think the former is more reliable. There is no API guarantee in our code or in miekg/dns that non-empty RRSets result in non-empty string representations.

Reply:
Hmm, but relying on the former requires plumbing the full set of records (i.e. including any iodef and any unrecognized ones) through to here. That's why I went with checking the raw string: our own code ensures that [the response string will be set if there are any Answers in the response](https://github.com/letsencrypt/boulder/blob/b7d9f8c2e32bf4ff91cb50553320e33acf178539/bdns/dns.go#L544-L553), and the miekg/dns .String() method [will never return an empty string](https://github.com/letsencrypt/boulder/blob/b7d9f8c2e32bf4ff91cb50553320e33acf178539/vendor/github.com/miekg/dns/msg.go#L900).

It seems to me that "dig-like output" is a pretty decent API guarantee for "if there were Answers, we'll print them".

Reply:
Okay, I've replaced this check with one against a new bool `caaRecord.present`, which is populated if and only if _any_ records are returned by the query.